doc/manual/packages.rst

```diff
@@ -346,6 +346,147 @@ Once you do this, the package manager knows your GitHub user name and can config
+  you can request that a new official release be tagged so that the
+  fix gets distributed to the rest of the community.) If you receive
+  an error ``Foo is dirty, bailing``, see :ref:`Dirty packages
+  <man-pkg-dirty>` below.
+
+- Create a branch for your changes: from the command line (not within
+  Julia), navigate to the package folder (e.g.,
+  ``~/.julia/v0.4/Foo/``) and type ``git checkout -b <newbranch>``,
+  where ``<newbranch>`` might be some descriptive name (e.g.,
+  ``fixbar``). By creating a branch, you ensure that you can easily go
+  back and forth between your new work and the current ``master``
+  branch (see
+  `<http://git-scm.com/book/en/v2/Git-Branching-Branches-in-a-Nutshell>`_).
+
+  If you forget to do this step until after you've already made some
+  changes, don't worry: see :ref:`more detail about branching
+  <man-post-hoc-branching>` below.
+
+- Make your changes. Whether it's fixing a bug or adding new
+  functionality, in most cases your change should include updates to
+  both the ``src/`` and ``test/`` folders.  If you're fixing a bug,
+  add your minimal example demonstrating the bug (on the current code)
+  to the test suite; by contributing a test for the bug, you ensure
+  that the bug won't accidentally reappear at some later time due to
+  other changes.  If you're adding new functionality, creating tests
+  demonstrates to the package owner that you've made sure your code
+  works as intended.
+
+- Run the package's tests and make sure they pass.  Usually you can do
+  this by running a file probably called ``runtests.jl`` from inside
+  the ``test/`` folder, or use ``Pkg.test("Foo")``.
+
+- Commit your changes: see `<http://git-scm.com/book/en/v2/Git-Basics-Recording-Changes-to-the-Repository>`_.
+
+- Submit your changes: from within julia, type
+  ``Pkg.submit("Foo")``. This will push your changes to your GitHub
+  fork, creating it if it doesn't already exist.  Julia will then give
+  you a hyperlink; open that link, edit the message, and then click
+  "submit." At that point, the package owner will be notified of your
+  changes and may initiate discussion.
+
+- The package owner may make some suggestions for additional
+  improvements.  To respond to those suggestions, you can easily
+  update the pull request (this only works for changes that have not
+  already been merged; for merged pull requests, make new changes by
+  starting a new branch):
+
+  + If you've changed branches in the meantime, make sure you go back
+    to the same branch with ``git checkout fixbar`` (from the command
+    line) or ``Pkg.checkout("Foo", "fixbar")`` (from julia).
+
+  + As above, make your changes, run the tests, and commit your changes.
+
+  + From the command line, type ``git push``.  This will add your new
+    commit(s) to the same pull request; you should see them appear
+    automatically on the page holding the discussion of your pull
+    request.
```

Comment:
In the ideal situation where everything is clean, `git rebase -i master && git push --force` should be enough. For other cases, well... http://git-scm.com/book/en/v2/Git-Tools-Rewriting-History